Primeiro *Challenge* que participei da **Alura**.
* Data Inicial: 11/07/2023
* Data Final: XX/XX/XXXX
Ao longo do desafio, outros cursos foram realizados para auxiliar na realização das tarefas. Desse modo, o notebook apresentado é uma combinação dos estudos com o desenvolvimento dos *challenges*.

> **OBJETIVO:** Avaliar a evação dos clientes de uma empresa, para a tomada de decisão para fidelização dos clientes.

## SetUp

In [1]:
import pandas as pd

In [2]:
pd.set_option('display.max_columns', 100)

In [3]:
uri = 'https://raw.githubusercontent.com/sthemonica/alura-voz/main/Dados/Telco-Customer-Churn.json'
data = pd.read_json(
    path_or_buf = uri,
    orient = 'columns'
)

data.head()

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

## Conhecendo o DataFrame

Os dados estão em *JSON*, contendo dicionários dentro das colunas. 

### Verificando os Dicionários

In [4]:
data['customer'][0]

{'gender': 'Female',
 'SeniorCitizen': 0,
 'Partner': 'Yes',
 'Dependents': 'Yes',
 'tenure': 9}

In [5]:
data['phone'][0]

{'PhoneService': 'Yes', 'MultipleLines': 'No'}

In [6]:
data['internet'][0]

{'InternetService': 'DSL',
 'OnlineSecurity': 'No',
 'OnlineBackup': 'Yes',
 'DeviceProtection': 'No',
 'TechSupport': 'Yes',
 'StreamingTV': 'Yes',
 'StreamingMovies': 'No'}

In [7]:
data['account'][0]

{'Contract': 'One year',
 'PaperlessBilling': 'Yes',
 'PaymentMethod': 'Mailed check',
 'Charges': {'Monthly': 65.6, 'Total': '593.3'}}

### Transformando os dados JSON em Tabela

In [8]:
dados_customer = pd.json_normalize(data['customer'])
dados_customer.head()


gender  SeniorCitizen Partner Dependents  tenure
0  Female              0     Yes        Yes       9
1    Male              0      No         No       9
2    Male              0      No         No       4
3    Male              1     Yes         No      13
4  Female              1     Yes         No       3

In [9]:
dados_phone = pd.json_normalize(data['phone'])
dados_phone.head()

PhoneService MultipleLines
0          Yes            No
1          Yes           Yes
2          Yes            No
3          Yes            No
4          Yes            No

In [10]:
dados_internet = pd.json_normalize(data['internet'])
dados_internet.head()

InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport  \
0             DSL             No          Yes               No         Yes   
1             DSL             No           No               No          No   
2     Fiber optic             No           No              Yes          No   
3     Fiber optic             No          Yes              Yes          No   
4     Fiber optic             No           No               No         Yes   

  StreamingTV StreamingMovies  
0         Yes              No  
1          No             Yes  
2          No              No  
3         Yes             Yes  
4         Yes              No

In [11]:
dados_account = pd.json_normalize(data['account'])
dados_account.head()

Contract PaperlessBilling     PaymentMethod  Charges.Monthly  \
0        One year              Yes      Mailed check             65.6   
1  Month-to-month               No      Mailed check             59.9   
2  Month-to-month              Yes  Electronic check             73.9   
3  Month-to-month              Yes  Electronic check             98.0   
4  Month-to-month              Yes      Mailed check             83.9   

  Charges.Total  
0         593.3  
1         542.4  
2        280.85  
3       1237.85  
4         267.4

### Concatenando as Tabelas

In [12]:
df = data[['customerID','Churn']]

df = pd.concat([df, dados_customer, dados_phone, dados_internet, dados_account], axis = 1)

print('O DataFrame original contém %d linhas.' % data.shape[0])
print('O DataFrame concatenado contém %d linhas.' % df.shape[0])

O DataFrame original contém 7267 linhas.
O DataFrame concatenado contém 7267 linhas.


### Analisando o DataFrame

In [13]:
df.head()

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService OnlineSecurity OnlineBackup  \
0          Yes            No             DSL             No          Yes   
1          Yes           Yes             DSL             No           No   
2          Yes            No     Fiber optic             No           No   
3          Yes            No     Fiber optic             No          Yes   
4          Yes            No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod  Charges.Monthly Charges.Total  
0              Yes      Mailed check             65.6         593.3  
1               No      Mailed check             59.9         542.4  
2              Yes  Electronic check             73.9        280.85  
3              Yes  Electronic check             98.0       1237.85  
4              Yes      Mailed check             83.9         267.4

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


A coluna *Charges.Total* está como objeto... <p>
A primeira tratativa foi utilizando o *.astype(float)*, mas o comando estava retornando um erro... Desse modo, outra ação foi tomada:

In [15]:
df['Charges Total'] = pd.to_numeric(df['Charges.Total'], errors='coerce')

df['Charges Total'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7267 entries, 0 to 7266
Series name: Charges Total
Non-Null Count  Dtype  
--------------  -----  
7256 non-null   float64
dtypes: float64(1)
memory usage: 56.9 KB


Com essa análise, percebeu-se que, das 7267 entradas, apenas 7256 continham informações que poderiam ser transformadas em float. <p>
Para compreender melhor, esses dados foram inseridos em uma nova coluna, filtrados e visualizados no próximo dataframe.

In [16]:
df[df['Charges Total'].isnull()]

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
975   1371-DWPAZ    No  Female              0     Yes        Yes       0   
1775  2520-SGTTA    No  Female              0     Yes        Yes       0   
1955  2775-SEFEE    No    Male              0      No        Yes       0   
2075  2923-ARZLG    No    Male              0     Yes        Yes       0   
2232  3115-CZMZD    No    Male              0      No        Yes       0   
2308  3213-VVOLG    No    Male              0     Yes        Yes       0   
2930  4075-WKNIU    No  Female              0     Yes        Yes       0   
3134  4367-NUYAO    No    Male              0     Yes        Yes       0   
3203  4472-LVYGI    No  Female              0     Yes        Yes       0   
4169  5709-LVOEQ    No  Female              0     Yes        Yes       0   
5599  7644-OMVMY    No    Male              0     Yes        Yes       0   

     PhoneService     MultipleLines InternetService       OnlineSecurity  \
975            No  No phone service             DSL                  Yes   
1775          Yes                No              No  No internet service   
1955          Yes               Yes             DSL                  Yes   
2075          Yes                No              No  No internet service   
2232          Yes                No              No  No internet service   
2308          Yes               Yes              No  No internet service   
2930          Yes               Yes             DSL                   No   
3134          Yes               Yes              No  No internet service   
3203           No  No phone service             DSL                  Yes   
4169          Yes                No             DSL                  Yes   
5599          Yes                No              No  No internet service   

             OnlineBackup     DeviceProtection          TechSupport  \
975                   Yes                  Yes                  Yes   
1775  No internet service  No internet service  No internet service   
1955                  Yes                   No                  Yes   
2075  No internet service  No internet service  No internet service   
2232  No internet service  No internet service  No internet service   
2308  No internet service  No internet service  No internet service   
2930                  Yes                  Yes                  Yes   
3134  No internet service  No internet service  No internet service   
3203                   No                  Yes                  Yes   
4169                  Yes                  Yes                   No   
5599  No internet service  No internet service  No internet service   

              StreamingTV      StreamingMovies  Contract PaperlessBilling  \
975                   Yes                   No  Two year               No   
1775  No internet service  No internet service  Two year               No   
1955                   No                   No  Two year              Yes   
2075  No internet service  No internet service  One year              Yes   
2232  No internet service  No internet service  Two year               No   
2308  No internet service  No internet service  Two year               No   
2930                  Yes                   No  Two year               No   
3134  No internet service  No internet service  Two year               No   
3203                  Yes                   No  Two year              Yes   
4169                  Yes                  Yes  Two year               No   
5599  No internet service  No internet service  Two year               No   

                  PaymentMethod  Charges.Monthly Charges.Total  Charges Total  
975     Credit card (automatic)            56.05                          NaN  
1775               Mailed check            20.00                          NaN  
1955  Bank transfer (automatic)            61.90                          NaN  
2075               Mailed check            19.70                          NaN  
22

Como é possível verificar, as informações não numéricas estavam vazias, mas não apareciam como *NaN* no DataFrame. <p>
Por serem apenas 11 informações, optou-se por excluí-los da base de dados.

In [17]:
df = df.dropna()
df['Charges.Total'] = df['Charges Total']
df.drop('Charges Total', axis = 1, inplace=True)
df.head()

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService OnlineSecurity OnlineBackup  \
0          Yes            No             DSL             No          Yes   
1          Yes           Yes             DSL             No           No   
2          Yes            No     Fiber optic             No           No   
3          Yes            No     Fiber optic             No          Yes   
4          Yes            No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod  Charges.Monthly  Charges.Total  
0              Yes      Mailed check             65.6         593.30  
1               No      Mailed check             59.9         542.40  
2              Yes  Electronic check             73.9         280.85  
3              Yes  Electronic check             98.0        1237.85  
4              Yes      Mailed check             83.9         267.40

In [18]:
df.describe()

SeniorCitizen       tenure  Charges.Monthly  Charges.Total
count      7256.0000  7256.000000      7256.000000    7256.000000
mean          0.1629    32.395535        64.755423    2280.634213
std           0.3693    24.558067        30.125739    2268.632997
min           0.0000     1.000000        18.250000      18.800000
25%           0.0000     9.000000        35.450000     400.225000
50%           0.0000    29.000000        70.300000    1391.000000
75%           0.0000    55.000000        89.900000    3785.300000
max           1.0000    72.000000       118.750000    8684.800000

### Análise do Conteúdo do DataFrame

In [19]:
for coluna in df.columns:
    print('*******')
    print(coluna)
    print(df[coluna].value_counts())

*******
customerID
0002-ORFBO    1
6615-NGGZJ    1
6646-JPPHA    1
6645-MXQJT    1
6641-XRPSU    1
             ..
3378-AJRAO    1
3376-BMGFE    1
3374-TTZTK    1
3374-PZLXD    1
9995-HOTOH    1
Name: customerID, Length: 7256, dtype: int64
*******
Churn
No     5163
Yes    1869
        224
Name: Churn, dtype: int64
*******
gender
Male      3669
Female    3587
Name: gender, dtype: int64
*******
SeniorCitizen
0    6074
1    1182
Name: SeniorCitizen, dtype: int64
*******
Partner
No     3747
Yes    3509
Name: Partner, dtype: int64
*******
Dependents
No     5086
Yes    2170
Name: Dependents, dtype: int64
*******
tenure
1     634
72    369
2     246
3     207
4     185
     ... 
28     60
38     60
39     59
44     54
36     50
Name: tenure, Length: 72, dtype: int64
*******
PhoneService
Yes    6551
No      705
Name: PhoneService, dtype: int64
*******
MultipleLines
No                  3490
Yes                 3061
No phone service     705
Name: MultipleLines, dtype: int64
*******
InternetServi

### Analisando o Churn

In [29]:
df['Churn'].isnull().sum()

0

In [30]:
df['Churn'].unique()

array(['No', 'Yes', ''], dtype=object)

In [36]:
porcent = len(df[df['Churn']=='']) / len(df) * 100
porcent

3.0871003307607494

Embora não apresente valores nulos, apresenta valores vazios. Optou-se por excluir esses dados da base de dados, por representarem 3% dos dados.

In [38]:
df = df[df['Churn']!='']

### Salvando o DataFrame

In [39]:
df.to_csv('dados_tratados.csv', index=False)